optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyRandomFactory.java

```diff
@@ -0,0 +1,107 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.heuristic.selector.common.nearby;
+
+import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
+
+import org.optaplanner.core.config.heuristic.selector.common.nearby.NearbySelectionConfig;
+import org.optaplanner.core.config.heuristic.selector.common.nearby.NearbySelectionDistributionType;
+
+public class NearbyRandomFactory {
+
+    public static NearbyRandomFactory create(NearbySelectionConfig nearbySelectionConfig) {
+        return new NearbyRandomFactory(nearbySelectionConfig);
+    }
+
+    private final NearbySelectionConfig nearbySelectionConfig;
+
+    public NearbyRandomFactory(NearbySelectionConfig nearbySelectionConfig) {
+        this.nearbySelectionConfig = nearbySelectionConfig;
+    }
+
+    public NearbyRandom buildNearbyRandom(boolean randomSelection) {
+        boolean blockDistributionEnabled =
+                nearbySelectionConfig.getNearbySelectionDistributionType() == NearbySelectionDistributionType.BLOCK_DISTRIBUTION
+                        || nearbySelectionConfig.getBlockDistributionSizeMinimum() != null
+                        || nearbySelectionConfig.getBlockDistributionSizeMaximum() != null
+                        || nearbySelectionConfig.getBlockDistributionSizeRatio() != null
+                        || nearbySelectionConfig.getBlockDistributionUniformDistributionProbability() != null;
+        boolean linearDistributionEnabled = nearbySelectionConfig
+                .getNearbySelectionDistributionType() == NearbySelectionDistributionType.LINEAR_DISTRIBUTION
+                || nearbySelectionConfig.getLinearDistributionSizeMaximum() != null;
+        boolean parabolicDistributionEnabled = nearbySelectionConfig
+                .getNearbySelectionDistributionType() == NearbySelectionDistributionType.PARABOLIC_DISTRIBUTION
+                || nearbySelectionConfig.getParabolicDistributionSizeMaximum() != null;
+        boolean betaDistributionEnabled =
+                nearbySelectionConfig.getNearbySelectionDistributionType() == NearbySelectionDistributionType.BETA_DISTRIBUTION
+                        || nearbySelectionConfig.getBetaDistributionAlpha() != null
+                        || nearbySelectionConfig.getBetaDistributionBeta() != null;
+        if (!randomSelection) {
+            if (blockDistributionEnabled || linearDistributionEnabled || parabolicDistributionEnabled
+                    || betaDistributionEnabled) {
+                throw new IllegalArgumentException("The nearbySelectorConfig (" + nearbySelectionConfig
+                        + ") with randomSelection (" + randomSelection
+                        + ") has distribution parameters.");
+            }
+            return null;
+        }
+        if (blockDistributionEnabled && linearDistributionEnabled) {
```

Comment:
Implementing my previous comment would complicate this. But I suppose you could set these variables in the `switch` directly.